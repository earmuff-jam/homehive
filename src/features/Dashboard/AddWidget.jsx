import React from "react";

import {
  Divider,
  ListItemIcon,
  ListItemText,
  MenuItem,
  MenuList,
  Stack,
  Typography,
} from "@mui/material";
import { AddRounded } from "@mui/icons-material";

const WidgetTypeList = [
  {
    id: 1,
    label: "Details Chart",
    inset: false, // makes text have extra spacing infront
  },
  {
    id: 2,
    label: "Invoice Details",
    inset: false,
  },
  {
    id: 3,
    label: "Invoice Demographics",
    inset: false,
  },
];

export default function AddWidget() {
  return (
    <Stack sx={{ minWidth: "12rem", padding: "1rem" }}>
      <Typography> Add Widget</Typography>
      <Divider />
      <MenuList dense>
        {WidgetTypeList.map((widgetType) => (
          <MenuItem key={widgetType.id}>
            <ListItemIcon>
              <AddRounded />
            </ListItemIcon>
            <ListItemText inset={widgetType.inset}>
              {widgetType.label}
            </ListItemText>
          </MenuItem>
        ))}
      </MenuList>
    </Stack>
  );
}
