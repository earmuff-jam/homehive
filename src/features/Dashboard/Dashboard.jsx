import { Button, IconButton, Stack, Tooltip } from "@mui/material";

import { AddRounded, RestartAltRounded } from "@mui/icons-material";

import RowHeader from "src/common/RowHeader/RowHeader";

export default function Dashboard() {
  return (
    <Stack>
      <Stack direction="row" justifyContent="space-between" alignItems="center">
        <RowHeader
          sxProps={{ textAlign: "left" }}
          title="Dashboard layout"
          caption="Displaying 3 out of 5 widgets"
        />
        <Stack direction="row" spacing={2}>
          <Tooltip title="Add Widget">
            <IconButton>
              <AddRounded fontSize="small" />
            </IconButton>
          </Tooltip>
          <Button variant="outlined" endIcon={<RestartAltRounded />}>
            Reset
          </Button>
        </Stack>
      </Stack>
    </Stack>
  );
}
