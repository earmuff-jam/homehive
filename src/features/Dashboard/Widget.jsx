import { CSS } from "@dnd-kit/utilities";
import { useSortable } from "@dnd-kit/sortable";

import { CancelRounded, DragIndicatorRounded } from "@mui/icons-material";
import { Badge, Box, IconButton, Paper, Stack, Tooltip } from "@mui/material";

import RowHeader from "src/common/RowHeader/RowHeader";

export default function Widget({ editMode, widget = {}, handleRemoveWidget }) {
  const { attributes, listeners, setNodeRef, transform, transition } =
    useSortable({
      id: widget.widgetID,
    });

  const style = {
    transition,
    transform: CSS.Transform.toString(transform),
  };

  return (
    <Box sx={{ ...style }} ref={setNodeRef}>
      <Badge
        badgeContent={
          editMode && (
            <IconButton
              size="small"
              color="error"
              disableRipple
              disableFocusRipple
              disableTouchRipple
              onClick={() => handleRemoveWidget(widget?.widgetID || "")}
            >
              <CancelRounded fontSize="small" />
            </IconButton>
          )
        }
      >
        <Box
          component={Paper}
          {...widget.config}
          sx={{
            padding: 1,
            overflow: "auto",
            backgroundColor: "background.paper",
          }}
        >
          <Stack direction="row" spacing={1}>
            {editMode && (
              <Tooltip title="Drag and drop to restructure widget layout">
                <IconButton
                  size="small"
                  {...attributes}
                  {...listeners}
                  disableRipple
                  disableFocusRipple
                  disableTouchRipple
                  color="primary"
                  sx={{ cursor: "move" }}
                >
                  <DragIndicatorRounded fontSize="inherit" />
                </IconButton>
              </Tooltip>
            )}
            <RowHeader
              title={widget.label}
              caption={widget.caption}
              sxProps={{
                textAlign: "left",
                fontWeight: "bold",
                color: "text.secondary",
              }}
            />
          </Stack>
        </Box>
      </Badge>
    </Box>
  );
}
