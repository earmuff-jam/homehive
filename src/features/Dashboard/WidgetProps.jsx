import { Box } from "@mui/material";
import InvoiceTimelineChart from "src/features/Widgets/InvoiceTimelineChart";
import InvoiceTrendsChart from "src/features/Widgets/InvoiceTrends";
import ItemTypeFreqChart from "src/features/Widgets/ItemTypeFreqChart";

export default function WidgetProps(widget) {
  switch (widget.id) {
    case 1:
      return (
        <Box sx={{ height: "20rem", width: "40rem" }}>
          <InvoiceTimelineChart label={widget.label} caption={widget.caption} />
        </Box>
      );
    case 2:
      return (
        <Box sx={{ height: "20rem", width: "40rem" }}>
          <InvoiceTrendsChart
            label={widget.label}
            caption={widget.caption}
          />
        </Box>
      );
    case 3:
      return (
        <Box sx={{ height: "20rem", width: "40rem" }}>
          <ItemTypeFreqChart
            label={widget.label}
            caption={widget.caption}
          />
        </Box>
      );
    default:
      return null;
  }
}
