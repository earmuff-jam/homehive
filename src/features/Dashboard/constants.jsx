export const WidgetTypeList = [
  {
    id: 1,
    label: "Invoice Tax Amount",
    caption: "Details about taxes from invoice list.",
    columns: [
      {
        label: "Label",
        accessorKey: "label",
      },
    ],
    data: [],
    config: {
      inset: false, // makes text have extra spacing infront
      height: "24rem",
      width: "48rem",
      minHeight: "24rem",
      minWidth: "48rem",
      widgetID: "9caef12d-a611-4573-8fd2-b5bd3036ce13", // widgetID for config is for provision only
    },
  },
  {
    id: 2,
    label: "Invoice Details",
    caption: "Details about your invoice.",
    columns: [],
    data: [],
    config: {
      inset: false,
      height: "24rem",
      width: "64rem",
      minHeight: "24rem",
      minWidth: "64rem",
      widgetID: "c04637c7-080d-4641-a4f4-4fd523280d74",
    },
  },
  {
    id: 3,
    label: "Invoice Demographics",
    caption:
      "Details about various recievers and their associated frequencies.",
    columns: [],
    data: [],
    config: {
      inset: false,
      height: "24rem",
      width: "24rem",
      minHeight: "24rem",
      minWidth: "24rem",
      widgetID: "052fda00-2d37-4d0f-81b7-3fcb451e5ee1",
    },
  },
];
