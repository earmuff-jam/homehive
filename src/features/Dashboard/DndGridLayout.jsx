import { useState } from "react";
import { Responsive as ResponsiveGridLayout } from "react-grid-layout";

import { Stack, Typography } from "@mui/material";
import Widget from "src/features/Dashboard/Widget";

export default function DndGridLayout({
  widgets,
  setWidgets,
  handleRemoveWidget,
}) {
  const [activeWidget, setActiveWidget] = useState(null); // active widget for drag overlay

  const handleDragStart = (ev) => {
    const { active } = ev;
    const activeId = active.id.toString();
    const widget = widgets.find((w) => w.widgetID === activeId);
    setActiveWidget(widget);
  };

  const handleDragEnd = (ev) => {
    const { active, over } = ev; // active is current, over is replacing

    setActiveWidget(null);

    if (!over) return;
    if (active.id === over.id) return;

    const originalIdx = widgets.findIndex(
      (widget) => widget.widgetID === active.id.toString()
    );
    const newIdx = widgets.findIndex(
      (widget) => widget.widgetID === over.id.toString()
    );

    const updatedWidgets = arrayMove(widgets, originalIdx, newIdx);
    setWidgets(updatedWidgets);
  };

  if (widgets.length <= 0)
    return (
      <Stack textAlign="center">
        <Typography variant="h5" sx={{ textTransform: "initial" }}>
          Sorry, no matching records found
        </Typography>
        <Typography variant="caption" sx={{ textTransform: "initial" }}>
          Add widgets for custom dashboard layout.
        </Typography>
      </Stack>
    );

  console.log(widgets);

  // const layout = [
  //   { widgetID: 1, x: 0, y: 0, w: 1, h: 2 },
  //   { widgetID: 2, x: 1, y: 0, w: 3, h: 2, minW: 2, maxW: 4 },
  //   { widgetID: 3, x: 4, y: 0, w: 1, h: 2 },
  // ];

  const baseLayout = widgets.map((widget, index) => ({
    i: widget.widgetID,
    x: (index % 4) * 12, // 4 per row, spread horizontally
    y: Math.floor(index / 4) * 8, // move vertically after 4 items
    w: 12,
    h: 8,
  }));

  const layouts = {
    lg: baseLayout,
    md: baseLayout,
    sm: baseLayout,
    xs: baseLayout,
    xxs: baseLayout,
  };

  return (
    <>
      <ResponsiveGridLayout
        margin={[20, 20]}
        containerPadding={[10, 10]}
        rowHeight={20}
        layouts={layouts}
        breakpoints={{ lg: 1200, md: 996, sm: 768, xs: 480, xxs: 0 }}
        cols={{ lg: 48, md: 48, sm: 48, xs: 48, xxs: 48 }}
        onLayoutChange={() => {
          console.log("layout has been changed. save it");
        }}
      >
        {widgets.map((widget) => (
          <div key={widget.widgetID}>
            <Widget widget={widget} handleRemoveWidget={handleRemoveWidget} />
          </div>
        ))}
      </ResponsiveGridLayout>
    </>
  );
}
