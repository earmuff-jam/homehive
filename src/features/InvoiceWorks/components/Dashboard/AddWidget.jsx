import React from "react";

import { AddRounded } from "@mui/icons-material";
import {
  Divider,
  ListItemIcon,
  ListItemText,
  MenuItem,
  MenuList,
  Stack,
  Typography,
} from "@mui/material";
import { WidgetTypeList } from "features/InvoiceWorks/components/Dashboard/constants";

export default function AddWidget({ handleAddWidget }) {
  return (
    <Stack sx={{ minWidth: "12rem", padding: "1rem" }}>
      <Typography> Add Widget</Typography>
      <Divider />
      <MenuList dense>
        {WidgetTypeList.map((widgetType) => (
          <MenuItem
            key={widgetType.id}
            onClick={() => handleAddWidget(widgetType.id)}
          >
            <ListItemIcon>
              <AddRounded />
            </ListItemIcon>
            <ListItemText inset={widgetType.config.inset}>
              {widgetType.label}
            </ListItemText>
          </MenuItem>
        ))}
      </MenuList>
    </Stack>
  );
}
