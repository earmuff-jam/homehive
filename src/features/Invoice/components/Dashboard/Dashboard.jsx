import React, { useEffect, useState } from "react";

import { v4 as uuidv4 } from "uuid";

import {
  AddRounded,
  EditRounded,
  RestartAltRounded,
} from "@mui/icons-material";
import {
  Box,
  Button,
  IconButton,
  Popover,
  Stack,
  Tooltip,
} from "@mui/material";
import CustomSnackbar from "common/CustomSnackbar/CustomSnackbar";
import RowHeader from "common/RowHeader/RowHeader";
import { pluralize } from "common/utils";
import AddWidget from "features/Invoice/components/AddWidget/AddWidget";
import DndGridLayout from "features/Invoice/components/DndGridLayout/DndGridLayout";
import { WidgetTypeList } from "features/Invoice/constants";
import { useAppTitle } from "hooks/useAppTitle";

export default function Dashboard() {
  useAppTitle("Dashboard");
  const [widgets, setWidgets] = useState([]);
  const [anchorEl, setAnchorEl] = useState(null);
  const [editMode, setEditMode] = useState(false); // re-arrange widgets
  const [showSnackbar, setShowSnackbar] = useState(false);

  const handleClose = () => setAnchorEl(null);
  const handleClick = (ev) => setAnchorEl(ev.currentTarget);

  const handleAddWidget = (id) => {
    const selectedWidget = WidgetTypeList.find(
      (widgetType) => widgetType.id === id,
    );

    setWidgets((prevWidgets) => {
      const updatedWidgets = [
        ...prevWidgets,
        // create a custom uuid to associate the widget for ui.
        // widgetID !== config.widgetID
        { ...selectedWidget, widgetID: uuidv4() },
      ];
      localStorage.setItem("widgets", JSON.stringify(updatedWidgets));

      return updatedWidgets;
    });

    setShowSnackbar(true);
    handleClose();
  };

  const handleRemoveWidget = (widgetID) => {
    setWidgets((prevWidgets) => {
      const remainingWidgets = prevWidgets.filter(
        (widget) => widget.widgetID !== widgetID,
      );

      localStorage.setItem("widgets", JSON.stringify(remainingWidgets));
      return remainingWidgets;
    });
  };

  const reset = () => {
    setWidgets([]);
    localStorage.setItem("widgets", JSON.stringify([]));
    handleClose();
  };

  useEffect(() => {
    const draftWidgets = localStorage.getItem("widgets");
    if (draftWidgets) {
      setWidgets(JSON.parse(draftWidgets));
    }
  }, []);

  return (
    <Stack>
      <Stack
        direction="row"
        justifyContent="space-between"
        alignItems="center"
        data-tour={"dashboard-0"}
      >
        <Stack direction="row">
          <Tooltip
            title={editMode ? "Save widget layout" : "Edit widget layout"}
          >
            <Box component="span">
              <IconButton
                sx={{ alignSelf: "flex-start" }}
                color="primary"
                disabled={widgets.length <= 0}
                onClick={() => setEditMode(!editMode)}
              >
                <EditRounded
                  fontSize="inherit"
                  color={editMode ? "primary" : "inherit"}
                />
              </IconButton>
            </Box>
          </Tooltip>
          <RowHeader
            sxProps={{ textAlign: "left", fontWeight: "bold" }}
            title={editMode ? "Editing Layout" : "Standard layout"}
            caption={`Displaying ${widgets.length} ${pluralize(
              widgets?.length,
              "widget",
            )}`}
          />
        </Stack>
        <Stack direction="row" spacing={2}>
          <Tooltip title="Add Widget">
            <IconButton onClick={handleClick} data-tour={"dashboard-1"}>
              <AddRounded fontSize="small" color="primary" />
            </IconButton>
          </Tooltip>
          <Button
            data-tour={"dashboard-2"}
            variant="outlined"
            endIcon={<RestartAltRounded />}
            onClick={reset}
          >
            Reset
          </Button>
        </Stack>
      </Stack>

      <Box data-tour={"dashboard-3"}>
        <DndGridLayout
          editMode={editMode}
          widgets={widgets}
          setWidgets={setWidgets}
          handleRemoveWidget={handleRemoveWidget}
        />
      </Box>

      {/* Add Widget Popover Content */}
      <Popover
        id={open ? "simple-popover" : undefined}
        open={Boolean(anchorEl)}
        anchorEl={anchorEl}
        onClose={handleClose}
        anchorOrigin={{
          vertical: "bottom",
          horizontal: "left",
        }}
      >
        <AddWidget handleAddWidget={handleAddWidget} />
      </Popover>

      <CustomSnackbar
        showSnackbar={showSnackbar}
        setShowSnackbar={setShowSnackbar}
        title="Successfully added new widget."
      />
    </Stack>
  );
}
