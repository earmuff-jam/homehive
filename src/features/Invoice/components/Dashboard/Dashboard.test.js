import React from "react";
import { render, screen } from "@testing-library/react";
import Dashboard from "./Dashboard";

describe("Dashboard", () => {
  it("renders correctly and matches snapshot", () => {
    const { asFragment } = render(<Dashboard />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders the Reset button", () => {
    render(<Dashboard />);
    expect(screen.getByText(/Reset/i)).toBeInTheDocument();
  });

  it("renders Add Widget button", () => {
    render(<Dashboard />);
    expect(screen.getByRole("button", { name: /Add Widget/i })).toBeInTheDocument();
  });
});
