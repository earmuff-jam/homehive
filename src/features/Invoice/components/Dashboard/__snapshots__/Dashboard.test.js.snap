// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiStack-root css-nen11g-MuiStack-root"
  >
    <div
      class="MuiStack-root css-3y1137-MuiStack-root"
      data-tour="dashboard-0"
    >
      <div
        class="MuiStack-root css-m69qwo-MuiStack-root"
      >
        <span
          aria-label="Edit widget layout"
          class="MuiBox-root css-0"
          data-mui-internal-clone-element="true"
        >
          <button
            class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-1aozkzf-MuiButtonBase-root-MuiIconButton-root"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1ckov0h-MuiSvgIcon-root"
              data-testid="EditRoundedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 17.46v3.04c0 .28.22.5.5.5h3.04c.13 0 .26-.05.35-.15L17.81 9.94l-3.75-3.75L3.15 17.1q-.15.15-.15.36M20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
              />
            </svg>
          </button>
        </span>
        <div
          class="MuiStack-root css-b0cywb-MuiStack-root"
        >
          <h5
            class="MuiTypography-root MuiTypography-h5 css-wj3xde-MuiTypography-root"
          >
            Standard layout
          </h5>
          <h6
            class="MuiTypography-root MuiTypography-subtitle2 css-1ypmsg6-MuiTypography-root"
          >
            Displaying 0 widget
          </h6>
          <div
            class="MuiBox-root css-56sg73"
          />
        </div>
      </div>
      <div
        class="MuiStack-root css-pl8gqh-MuiStack-root"
      >
        <button
          aria-label="Add Widget"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-53g0n7-MuiButtonBase-root-MuiIconButton-root"
          data-mui-internal-clone-element="true"
          data-tour="dashboard-1"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeSmall css-os5i4q-MuiSvgIcon-root"
            data-testid="AddRoundedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M18 13h-5v5c0 .55-.45 1-1 1s-1-.45-1-1v-5H6c-.55 0-1-.45-1-1s.45-1 1-1h5V6c0-.55.45-1 1-1s1 .45 1 1v5h5c.55 0 1 .45 1 1s-.45 1-1 1"
            />
          </svg>
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1blvszr-MuiButtonBase-root-MuiButton-root"
          data-tour="dashboard-2"
          tabindex="0"
          type="button"
        >
          Reset
          <span
            class="MuiButton-icon MuiButton-endIcon MuiButton-iconSizeMedium css-1gulhci-MuiButton-endIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="RestartAltRoundedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 5V3.21c0-.45-.54-.67-.85-.35l-2.8 2.79c-.2.2-.2.51 0 .71l2.79 2.79c.32.31.86.09.86-.36V7c3.31 0 6 2.69 6 6 0 2.72-1.83 5.02-4.31 5.75-.42.12-.69.52-.69.95 0 .65.62 1.16 1.25.97C17.57 19.7 20 16.64 20 13c0-4.42-3.58-8-8-8m-6 8c0-1.34.44-2.58 1.19-3.59.3-.4.26-.95-.09-1.31-.42-.42-1.14-.38-1.5.1-1 1.34-1.6 3-1.6 4.8 0 3.64 2.43 6.7 5.75 7.67.63.19 1.25-.32 1.25-.97 0-.43-.27-.83-.69-.95C7.83 18.02 6 15.72 6 13"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
    <div
      class="MuiBox-root css-0"
      data-tour="dashboard-3"
    >
      <div
        class="MuiStack-root css-5andtu-MuiStack-root"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5 css-16xl4zq-MuiTypography-root"
        >
          Sorry, no matching records found
        </h5>
        <span
          class="MuiTypography-root MuiTypography-caption css-mpaqru-MuiTypography-root"
        >
          Add widgets for custom dashboard layout.
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
